lib/Mailer/SparkPost.php

```diff
@@ -0,0 +1,131 @@
+ */
+class SparkPost extends Mailer
+{
+    protected function getEndpoint()
+    {
+        return 'https://api.sparkpost.com/api/v1/transmissions';
+    }
+
+    protected function getHeaders()
+    {
+        return [
+            'Authorization' => $this->getServerToken(),
+        ];
+    }
+
+    protected function format(MessageInterface $message)
+    {
+        $from = $this->normalizeIdentity($message->getFrom());
+
+        $parameters = [
+            'options' => [
+                'transactional' => true,
+            ],
+            'content' => [
+                'from' => [
+                    'name' => $from->getName(),
+                    'email' => $from->getEmail(),
+                ],
+                'subject' => $message->getSubject(),
+                'text' => $message->getText(),
+                'html' => $message->getHtml(),
+            ],
+        ];
+
+        if ($message->getReplyTo()) {
+            $parameters['content']['reply_to'] = $message->getReplyTo();
+        }
+
+        if (count($message->getHeaders())) {
+            $parameters['content']['headers'] = $message->getHeaders();
+        }
+
+        if ($message instanceof AttachmentsAwareInterface) {
+            foreach ($message->getAttachments() as $attachment) {
+                $inline = $attachment->getId() !== null;
+                $parameters['content'][$inline ? 'inline_images' : 'attachments'][] = [
+                    'type' => $attachment->getType(),
+                    'name' => $inline ? $attachment->getId() : $attachment->getName(),
+                    'data' => base64_encode($this->getAttachmentContent($attachment)),
+                ];
+            }
+        }
+
+        if ($message instanceof MetadataAwareInterface) {
+            $metadata = array_filter($message->getMetadata());
+            if (count($metadata)) {
+                $parameters['metadata'] = $metadata;
+            }
+        }
+
+        $tags = [];
+        if ($message instanceof TaggableInterface) {
+            $tags = (array) $message->getTag();
+        }
+
+        $normalizedTo = $this->normalizeIdentities($message->getTo());
+        $normalizedCc = $this->normalizeIdentities($message->getCc());
+        $normalizedBcc = $this->normalizeIdentities($message->getBcc());
+        $toIdentityString = IdentityUtils::buildIdentityString($normalizedTo);
+
+        /** @var \Stampie\Identity[] $merged */
+        $merged = array_merge($normalizedTo, $normalizedCc, $normalizedBcc);
+        foreach ($merged as $recipient) {
+            /*
+             * IMPORTANT:
+             * Do not set address.name!
+             * SparkPost will wrap the entire header_to value, e.g. Bob <bob@example.com, john@example.com>
+             */
+
+            $parameters['recipients'][] = [
+                'address' => [
+                    'email' => $recipient->getEmail(),
+                    'header_to' => $toIdentityString,
```

Comment:
Reading the source of the official SparkPost Node SDK, this should be a bit different:

- for normal recipients, use `email` and `name` in the JSON recipients
- for CC and BCC, use `email` and `header_to`

Reply:
It's been a while, so I don't remember the details, but I remember having issues with how SparkPost was building the header. I'll take a look again when I get a chance.

Reply:
Looking at why I had it this way again. In the following scenarios, a message has the following "To" recipients: **john@example.com** & **sally@example.com**

# Scenario A:

```php
$parameters['recipients'] = [
  ['address' => ['email' => 'john@example.com', 'name' => 'John']],
  ['address' => ['email' => 'sally@example.com', 'name' => 'Sally']],
];
```

## Scenario A "To" headers

**John's Inbox** - `John <john@example.com>`
**Sally's Inbox** - `Sally <sally@example.com>`

This was undesired behavior for us. We wanted the To header to be `John <john@example.com>,Sally <sally@example.com>`. In other words, show everybody who received the email. If we wanted to truly mimic Scenario A's behavior, then we would loop each recipient, call `->setTo($recipient)` and send.

# Scenario B:

```php
$parameters['recipients'] = [
  ['address' => [
    'email' => 'john@example.com', 
    'header_to' => 'John <john@example.com>,Sally <sally@example.com>'
  ]],
  ['address' => [
    'email' => 'sally@example.com', 
    'header_to' => 'John <john@example.com>,Sally <sally@example.com>'
  ]],
];
```

## Scenario B "To" headers

**John's Inbox** - `John <john@example.com>,Sally <sally@example.com>`
**Sally's Inbox** - `John <john@example.com>,Sally <sally@example.com>`

This is what we wanted. Each recipient receives the email and is able to see everybody else that it was sent to in the To header.

## Why not set the 'name' key in Scenario B?

If `name` is provided, SparkPost mangles the To header like so:

`"John <John" <john@example.com>,"Sally >" <sally@example.com>`